modules/dxp/apps/akismet/akismet-web/src/main/resources/META-INF/resources/view.jsp

```diff
@@ -28,64 +28,54 @@ request.setAttribute("view.jsp-portletURL", portletURL);
 
 <%@ include file="/message_boards.jspf" %>
 
-<aui:script>
-	Liferay.provide(
-		window,
-		'<portlet:namespace />deleteMBMessages',
-		function (dicussion) {
-			var deleteMBMessageIds = Liferay.Util.listCheckedExcept(
-				document.<portlet:namespace />fm,
-				'<portlet:namespace />allRowIds'
-			);
+<aui:script use="liferay-util-list-fields">
```

Comment:
This adds race conditions and I don't see anything in the explanation in the PR to state that we've checked and it's okay as we did with others.

You've effectively moved from a synchronous `window.X` definition that was always available to an asynchronous one, so you should be able to guarantee that `window.X` can't be called before being defined.

Is that the case?

Reply:
I've tested this change.

I know is that the functions (`window['<portlet:namespace />deleteMBMessages']` and `window['<portlet:namespace />notSpamMBMessages']`) are called when clicking on these buttons if a _discussion_ is selected (if nothing is selected, these buttons don't do anything)

![delete-or-discard-buttons](https://user-images.githubusercontent.com/5572/84895874-4c086d80-b0a3-11ea-9fbd-36c493d89e50.png)

Reply:
> I've tested this change.

I believe you know this @julien, but just in case somebody else reads this, I just want to remind that testing is great, but when it comes to *races*, seeing it work once (or twice, or 100 times) isn't proof that it is always going to work. The only way you can approach certainty about that is to analyze the data dependencies, which I think you might be getting at here.

Are you saying that by the time a user can click these buttons (and therefore, by which time the functions must be defined) that `liferay-util-list-fields` must already have loaded, because otherwise we wouldn't even have shown the list yet and/or the user wouldn't have been able to make a selection?

If (and only if) that is the case, then it seems you could remove the `use` too. Can you confirm that? (Or alternatively, say why the `use` needs to stay there.) I've only looked at this in the most superficial way, so I don't have a deep understanding of it like you have.

Reply:
Yes, I'm saying that by the time the use clicks the buttons those two functions will be defined.

I might be too optimistic as usual but this is also one test I did 

![on-load](https://user-images.githubusercontent.com/5572/84899425-ef0fb600-b0a8-11ea-80c6-800c906d973d.png)

Reply:
Yeah, testing is helpful, but when it comes to race conditions, testing alone doesn't tell you about whether something happens before something else or not (that's why they're called races). eg. you might have `A` and `B` getting defined in some indeterminate order. Like a coin flip, you might check 1, 2, 3, or N times and always see `A` first. But doesn't mean there isn't a race, just that you didn't see `B` win the race.

In this case, and the case of any race, you really have to analyze the data dependencies. For example, in this case, `git grep liferay-util-list-fields` doesn't show that AUI module being defined anywhere... weird, huh? A search (ie. `git log -Gliferay-util-list-fields -p`) reveals c098e6b3d72313bb621c5ef349af802d62a9eb9f with a pretty interesting message:

> since liferay-util-list-fields module was removed; No need for Liferay.provide()

`git grep listCheckExcept` shows it being defined [here](https://github.com/liferay/liferay-portal/blob/39c5030e7d18fe2f3ee95db5b612b344c8ba0990/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/util.js#L621), and I haven't found the commit yet that moved it there, but from this we can already conclude that the `use` is probably bogus, and all we appear to have a dependency on the base AUI `Util` class instead.

Reply:
Took a while to find it, but moved in efd9c67e18b4536912f3e9ecef64987271304529 ("[LPS-51317](https://issues.liferay.com/browse/LPS-51317) Remove util_list_fields.js", November 2013).

Reply:
Thanks, I was still searching for that.
I'll remove that `use` and see what happens.

![](https://media.giphy.com/media/XUFPGrX5Zis6Y/giphy.gif)

Reply:
Please don't tell me that "see what happens" means another manual test...

That's not the way to verify that we don't have a race bug.

What we know:

- If you remove the `use` (which does nothing, I guess 🤷‍♂️, because `liferay-util-list-fields` doesn't exist) then the functions *will* be eagerly defined. Good.
- If you click on the button before `Liferay.Util.listCheckedExcept` is defined, you will indeed get an explosion.
- `Liferay.Util.listCheckedExcept` is defined as part of the `liferay-util` AUI module, so that's what the real dependency is.
- `liferay-util` is bootstrapped quite early on in the page lifecycle because it is [a "core" module](https://github.com/liferay/liferay-portal/blob/39c5030e7d18fe2f3ee95db5b612b344c8ba0990/modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/servlet/AUITopHeadResources.java#L93).

From all this, I conclude, there is no race, but there is a dependency which will block us from turning off frontend-js-aui-web by default. At least, until we address [LPS-101986](https://issues.liferay.com/browse/LPS-101986).

I hope this is helpful as an example of the analysis that we should be doing on all of these pulls, and reflecting in the description. It can save a bit of back and forth (and also duplicated effort, where I verify things that you have probably already done).